composeApp/src/commonMain/kotlin/io/github/couchtracker/db/user/datastruct/partialtime/PartialDateTime.kt

```diff
@@ -0,0 +1,471 @@
+                is YearMonth -> local.month.value
+                is Date -> local.date.month.value
+                is DateTime -> local.dateTime.month.value
+            }
+        }
+
+        /**
+         * Sorts [PartialDateTime] only by their local part. At parity of local part, [Zoned] values are put last.
+         */
+        private val LOCAL_COMPARATOR: Comparator<PartialDateTime> = LOCAL_YEAR_MONTH_COMPARATOR.thenBy {
+            when (val local = it.local) {
+                is Year, is YearMonth -> Int.MIN_VALUE
+                is Date -> local.date.toEpochDays()
+                is DateTime -> local.dateTime.date.toEpochDays()
+            }
+        }.thenBy {
+            when (val local = it.local) {
+                is Year, is YearMonth, is Date -> Instant.DISTANT_PAST
+                is DateTime -> local.dateTime.toInstant(UtcOffset.ZERO)
+            }
+        }.thenBy {
+            when (it) {
+                is Local -> 0
+                is Zoned -> 1
+            }
+        }
+
+        private val ZONED_COMPARATOR: Comparator<Zoned> = compareBy<Zoned, Local>(LOCAL_YEAR_MONTH_COMPARATOR) { it.local }
+            .thenBy {
+                when (it.local) {
+                    is Year, is YearMonth -> 0
+                    is Date, is DateTime -> 1
+                }
+            }
+            .thenBy { it.toInstant() }
+            .thenBy {
+                when (it.local) {
+                    is Year -> 0
+                    is YearMonth -> 1
+                    is Date -> 2
+                    is DateTime -> 3
+                }
+            }
+
+        /**
+         * Sorts the given [items] based on their [PartialDateTime], retrieved with the [getPartialDateTime] lambda.
+         * Upon identical [PartialDateTime], [additionalComparator] is used on the items [T].
+         *
+         * The algorithm works like this:
+         * - First it splits [Local] and [Zoned] values into two different lists, and sorts the according to their comparator;
+         * - Then, the two lists are merged together, sorted according to their local part only.
+         *
+         * This ensures that the [Local] and [Zoned] are always sorted among themselves.
+         *
+         * @see PartialDateTime.sortAndGroup
+         */
+        fun <T> sort(
+            items: Iterable<T>,
+            getPartialDateTime: T.() -> PartialDateTime,
+            additionalComparator: Comparator<T> = compareBy { 0 },
+        ): List<T> {
+            data class ItemWithDate(val item: T, val date: PartialDateTime)
+
+            val itemsWithDates = items.map { ItemWithDate(item = it, date = it.getPartialDateTime()) }
+
+            val locals = itemsWithDates
+                .asSequence()
+                .filter { it.date is Local }
+                .sortedWith(compareBy<ItemWithDate, Local>(LOCAL_COMPARATOR) { it.date as Local }.thenBy(additionalComparator) { it.item })
+                .toCollection(ArrayDeque())
```

Comment:
Suggestion: instead of `ArrayDeque`, have you tried a normal list, and use an iterator instead of calling `removeFirst()` consecutively?